zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -772,104 +795,108 @@ where
         },
     );
 
+    #[cfg(feature = "transparent-inputs")]
+    let mut has_shielded_inputs = false;
+
     for (sapling_key, sapling_note, merkle_path) in sapling_inputs.into_iter() {
         builder.add_sapling_spend(&sapling_key, sapling_note.clone(), merkle_path)?;
+        #[cfg(feature = "transparent-inputs")]
+        {
+            has_shielded_inputs = true;
+        }
     }
 
     #[cfg(feature = "orchard")]
     for (orchard_note, merkle_path) in orchard_inputs.into_iter() {
         builder.add_orchard_spend(usk.orchard(), *orchard_note, merkle_path.into())?;
+        #[cfg(feature = "transparent-inputs")]
+        {
+            has_shielded_inputs = true;
+        }
     }
 
     #[cfg(feature = "transparent-inputs")]
-    let utxos_spent = {
-        let known_addrs = wallet_db
-            .get_transparent_receivers(account)
-            .map_err(Error::DataSource)?;
+    let mut known_addrs = wallet_db
+        .get_transparent_receivers(account_id)
+        .map_err(Error::DataSource)?;
+    #[cfg(feature = "transparent-inputs")]
+    let mut ephemeral_added = false;
+
+    #[cfg(feature = "transparent-inputs")]
+    let mut metadata_from_address = |addr: TransparentAddress| -> Result<
+        TransparentAddressMetadata,
+        ErrorT<DbT, InputsErrT, FeeRuleT>,
+    > {
+        match known_addrs.get(&addr) {
+            None if !ephemeral_added => {
+                // The ephemeral addresses are added lazily to avoid extra database operations
+                // in the common case. We don't need to include them in order to be able to
```

Comment:
What is the common case?

Reply:
That the user doesn't use TEX addresses at all.